sale_coupon_product_exclude/models/sale_order.py

```diff
@@ -0,0 +1,56 @@
+# Copyright 2022 Dinar Gabbasov
+# Copyright 2022 Ooops404
+# Copyright 2022 Cetmix
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def _get_base_order_lines(self, program):
+        order_lines = super(SaleOrder, self)._get_base_order_lines(program)
+        # exclude products from base order lines: the maximum discount cannot be greater than
+        # the total sum of the base order lines with excluded products
+        order_lines = order_lines.filtered(
+            lambda line: not program._is_excluded_product(line.product_id)
+        )
+        return order_lines
+
+    def _get_reward_values_discount(self, program):
+        # we need to run the original method with the updated context
+        # so that we can use the program to filter the order lines.
+        return super(
+            SaleOrder, self.with_context(coupon_program=program)
```

Comment:
Maybe this context key could be more specific to avoid potential collisions with other modules

Reply:
> Maybe this context key could be more specific to avoid potential collisions with other modules

Thank you for the review, I renamed "coupon_program" -> "current_coupon_program"